compiler/rustc_trait_selection/src/traits/specialize/mod.rs

```diff
@@ -595,3 +598,128 @@ fn report_conflicting_impls<'tcx>(
         }
     }
 }
+
+pub(super) fn trait_has_impl_which_may_shadow_dyn<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    (target_trait_def_id, principal_def_id): (DefId, Option<DefId>),
+) -> bool {
+    // We only care about trait objects which have associated types.
+    if !tcx
+        .associated_items(target_trait_def_id)
+        .in_definition_order()
+        .any(|item| item.kind == ty::AssocKind::Type)
+    {
+        return false;
+    }
+
+    let target_self_ty =
+        principal_def_id.map_or(SimplifiedType::MarkerTraitObject, SimplifiedType::Trait);
+
+    let elaborated_supertraits =
+        principal_def_id.into_iter().flat_map(|def_id| tcx.supertrait_def_ids(def_id)).collect();
+
+    trait_has_impl_inner(
+        tcx,
+        target_trait_def_id,
+        target_self_ty,
+        &elaborated_supertraits,
+        &mut Default::default(),
+        false,
+    )
+}
+
+fn trait_has_impl_inner<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    target_trait_def_id: DefId,
+    target_self_ty: SimplifiedType<DefId>,
+    elaborated_supertraits: &FxHashSet<DefId>,
+    seen_traits: &mut FxHashSet<DefId>,
+    for_nested_impl: bool,
```

Comment:
pls change to an `enum ForNestedImpl`

or wait: alternative:

if we duplicate the "is `Sized` check" in `fn trait_has_impl_which_may_shadow_dyn` and have it call an extracted function `fn check_user_written_impls`, which then recurse into `trait_has_impl_inner`, we don't need `ForNestedImpl` at all.